qiime/core/callable.py

```diff
@@ -0,0 +1,336 @@
+            parameters = {}
+            for name in self.signature.parameters:
+                parameters[name] = args[name]
+
+            view_args = parameters.copy()
+            for name, (_, view_type) in self.signature.inputs.items():
+                view_args[name] = artifacts[name].view(view_type)
+
+            execution_uuid = uuid.uuid4()
+            action_reference = (
+                "%s. Details on plugin, version, website, etc. will also be "
+                "included, see https://github.com/qiime2/qiime2/issues/26"
+                % self.id)
+            artifact_uuids = {name: artifact.uuid for name, artifact in
+                              artifacts.items()}
+            parameter_references = {name: str(param) for name, param in
+                                    parameters.items()}
+            provenance = qiime.sdk.Provenance(
+                execution_uuid, action_reference, artifact_uuids,
+                parameter_references)
+
+            outputs = self._callable_executor_(self._callable, view_args,
+                                               output_types, provenance)
+            # `outputs` matches a Python function's return: either a single
+            # value is returned, or it is a tuple of return values.
+            for output in tuplize(outputs):
+                output._orphan(self._pid)
+            return outputs
+
+        callable_wrapper = self._rewrite_wrapper_signature(callable_wrapper)
+        self._set_wrapper_properties(callable_wrapper, '__call__')
+        return callable_wrapper
+
+    def _get_async_wrapper(self):
+        def async_wrapper(*args, **kwargs):
+            # TODO handle this better in the future, but stop the massive error
+            # caused by MacOSX async runs for now.
+            try:
+                import matplotlib as plt
+                if plt.rcParams['backend'] == 'MacOSX':
+                    raise EnvironmentError(backend_error_template %
+                                           plt.matplotlib_fname())
+            except ImportError:
+                pass
+
+            args = args[1:]
+            pool = concurrent.futures.ProcessPoolExecutor(max_workers=1)
+            future = pool.submit(self, *args, **kwargs)
+            pool.shutdown(wait=False)
+            return future
+
+        async_wrapper = self._rewrite_wrapper_signature(async_wrapper)
+        self._set_wrapper_properties(async_wrapper, 'async')
+        return async_wrapper
+
+    def _rewrite_wrapper_signature(self, wrapper):
+        # Convert the callable's signature into the wrapper's signature and set
+        # it on the wrapper.
+        return decorator.decorator(
+            wrapper, self._callable_sig_converter_(self._callable))
+
+    def _set_wrapper_properties(self, wrapper, name):
+        wrapper.__name__ = wrapper.__qualname__ = name
+        wrapper.__module__ = self.get_import_path()
+        del wrapper.__annotations__
+        del wrapper.__wrapped__
```

Comment:
Comment this is necessary so that `inspect` doesn't display wrapped function's annotations